Truck route index

Betristor opened the pull request:
1. Change truck model from indexed by zone to indexed by route
2. Change truck compression into Comp
3. Write truck number on each route

Betristor commented:
@anna-nadia Hi, Anna. The truck route now is maintained in the file named 'Routes.csv' with three columns. The first and second column specify the start zone and end zone for one route. The direction is modeled in the model so here start and end is just for denoting a route wiout special meaning. The third column gives the distance of corresponding route.

anna-nadia reviewed (commented):
Hi Yuheng, thanks for the updates. There are many code changes so it's a bit difficult for me to review well without running a test. I can try to run a test on my model before our next DOLPHYN meeting. Here are some comments:
1) does it still make sense to have "Existing_Energy_Cap_tonne_zX" as inputs now that we are modeling trucks by route?
2)  although I see you made changes in the variable naming for compression, it's still unclear if you are modeling compression in tonnes or in tonnes/hour. Previously, I understood it was in tonnes (so energy units), but this is unusual for a compressor - which is usually modeled based on power. 
3) It would be good to update the sample results of the example system when you commit, since the input files are updated, and it would helpful to see what the new output files look like. 
4) Adding a few more comments to the code could be helpful for future users - for example, I'm not sure if the new "d" variable for direction is explained. 
It seems that my previous changes haven't been merged to the main branch yet, but they should be soon - which means that you may have a few conflicts to resolve - but I think they should be manageable. I just updated the H2 balance expression for liquid versus gas, and also added in OPEX costs. However, I did not write something explicit for LOHC in the balance expression, which could be an issue - and should probably be treated as GAS. Can do this later if needed.

dharik13 commented:
Yuheng will test updated truck module with Anna's data set and report back on performance. Once we can converge on the performance and model outputs, we can consider merging.

dharik13 commented:
Check units for compression variables and parameters.

RuaridhMacd commented:
- [ ] @Betristor Write a comment on the performance / runtime improvements based on the changes so far, so that @anna-nadia is up to speed.
- [ ] @RuaridhMacd Look over the code being used to write the outputs.

Betristor commented:
Considering the complexity caused by truck dispatch, the previous index way of constructing each potential route by enumerating the combination of all zones is abandoned. Instead, users could offer routes of interest in routes.csv by providing start zone and end zone as well as distances. 

After testing anna's case with new index way (ten faked routes are used, console output below, writing time is lost), it succeeded to solve the model and yield results, taking estiamted ten hours among which two hours for solving model, so now accessing the results becomes the most time-consuming part. 

In summary, the new truck index way helped reduce the time for establishing and solving model but obtaining results becomes unreasonably time-consuming.

```
Loading packages
[ Info: Precompiling DOLPHYN [top-level]
[ Info: PACKAGE LOADED
Time Series Data Already Clustered.
Configuring Solver
Reading Input CSV Files
[ Info: Reading Input CSV Files
Network.csv Successfully Read!
[ Info: Network.csv Successfully Read!
Load_data.csv Successfully Read!
[ Info: Load_data.csv Successfully Read!
Fuels_data.csv Successfully Read!
[ Info: Fuels_data.csv Successfully Read!
Generators_data.csv Successfully Read!
[ Info: Generators_data.csv Successfully Read!
Generators_variability.csv Successfully Read!
[ Info: Generators_variability.csv Successfully Read!
CO2_cap.csv Successfully Read!
[ Info: CO2_cap.csv Successfully Read!
CSV Files Successfully Read In From /home/zyh/DOLPHYN-dev/Example_Systems/Aviation_Liq_Truck/
[ Info: CSV Files Successfully Read In From /home/zyh/DOLPHYN-dev/Example_Systems/Aviation_Liq_Truck/
Reading H2 Input CSV Files
[ Info: Reading H2 Input CSV Files
HSC_generation.csv Successfully Read!
[ Info: HSC_generation.csv Successfully Read!
HSC_load_data.csv Successfully Read!
[ Info: HSC_load_data.csv Successfully Read!
HSC_generators_variability.csv Successfully Read!
[ Info: HSC_generators_variability.csv Successfully Read!
HSC_pipelines.csv Successfully Read!
[ Info: HSC_pipelines.csv Successfully Read!
Routes.csv Successfully Read!
[ Info: Routes.csv Successfully Read!
HSC_trucks.csv Successfully Read!
[ Info: HSC_trucks.csv Successfully Read!
HSC_g2p_variability.csv Successfully Read!
[ Info: HSC_g2p_variability.csv Successfully Read!
Period_map.csv Successfully Read!
[ Info: Period_map.csv Successfully Read!
HSC Input CSV Files Successfully Read In From /home/zyh/DOLPHYN-dev/Example_Systems/Aviation_Liq_Truck/
[ Info: HSC Input CSV Files Successfully Read In From /home/zyh/DOLPHYN-dev/Example_Systems/Aviation_Liq_Truck/
Set parameter Username
Academic license - for non-commercial use only - expires 2023-11-08
Set parameter Presolve to value 1
Set parameter AggFill to value 10
Set parameter PreDual to value 0
Set parameter TimeLimit to value 110000
Set parameter Method to value 2
Set parameter Crossover to value 0
Set parameter LogFile to value "gurobi.log"
Discharge Module
[ Info: Discharge Module
Non-served Energy Module
[ Info: Non-served Energy Module
Investment Discharge Module
[ Info: Investment Discharge Module
Unit Commitment Module
[ Info: Unit Commitment Module
Emissions Module for CO2 Policy modularization
[ Info: Emissions Module for CO2 Policy modularization
Transmission Module
[ Info: Transmission Module
Dispatchable Resources Module
[ Info: Dispatchable Resources Module
Must-Run Resources Module
[ Info: Must-Run Resources Module
Storage Resources Module
[ Info: Storage Resources Module
Storage Investment Module
[ Info: Storage Investment Module
Storage Core Resources Module
[ Info: Storage Core Resources Module
Storage Resources with Symmetric Charge/Discharge Capacity Module
[ Info: Storage Resources with Symmetric Charge/Discharge Capacity Module
Hydro Reservoir Core Resources Module
[ Info: Hydro Reservoir Core Resources Module
Thermal (Unit Commitment) Resources Module
[ Info: Thermal (Unit Commitment) Resources Module
Hydrogen Generation and Storage Discharge Module
[ Info: Hydrogen Generation and Storage Discharge Module
Hydrogen Investment Discharge Module
[ Info: Hydrogen Investment Discharge Module
Hydrogen Production Module
[ Info: Hydrogen Production Module
H2 Production Core Module
[ Info: H2 Production Core Module
H2 Production (No Unit Commitment) Module
[ Info: H2 Production (No Unit Commitment) Module
H2 Emissions Module for CO2 Policy modularization
[ Info: H2 Emissions Module for CO2 Policy modularization
Hydrogen Non-served Energy Module
[ Info: Hydrogen Non-served Energy Module
Hydrogen Storage Module
[ Info: Hydrogen Storage Module
H2 Storage Energy Investment Module
[ Info: H2 Storage Energy Investment Module
H2 Storage Charging Investment Module
[ Info: H2 Storage Charging Investment Module
H2 Storage Core Resources Module
[ Info: H2 Storage Core Resources Module
Hydrogen Long Duration Storage Module
[ Info: Hydrogen Long Duration Storage Module
┌ Warning: Axis contains one element: 365. If intended, you can safely ignore this warning. To explicitly pass the axis with one element, pass `[365]` instead of `365`.
└ @ JuMP.Containers ~/.julia/packages/JuMP/Z1pVn/src/Containers/DenseAxisArray.jl:173
Hydrogen Pipeline Module
[ Info: Hydrogen Pipeline Module
Hydrogen Truck Module
[ Info: Hydrogen Truck Module
H2 Truck Investment Module
[ Info: H2 Truck Investment Module
H2 Truck Core Resources Module
[ Info: H2 Truck Core Resources Module
H2 g2p demand module
[ Info: H2 g2p demand module
C02 Policies Module for power and hydrogen system combined
[ Info: C02 Policies Module for power and hydrogen system combined
Solving Model
Set parameter PreDual to value 0
Set parameter Method to value 2
Set parameter Crossover to value 0
Set parameter TimeLimit to value 110000
Set parameter LogFile to value "gurobi.log"
Set parameter AggFill to value 10
Set parameter Presolve to value 1
Gurobi Optimizer version 9.5.1 build v9.5.1rc2 (linux64)
Thread count: 28 physical cores, 56 logical processors, using up to 28 threads
Optimize a model with 10621124 rows, 5860856 columns and 35318021 nonzeros
Model fingerprint: 0x7a2879fe
Coefficient statistics:
 Matrix range [8e-05, 5e+03]
 Objective range [2e-06, 1e+04]
 Bounds range [0e+00, 0e+00]
 RHS range [3e-17, 1e+07]
Presolve removed 0 rows and 0 columns (presolve time = 14s) ...
Presolve removed 1235650 rows and 118163 columns (presolve time = 16s) ...
Presolve removed 1272931 rows and 537430 columns (presolve time = 33s) ...
Presolve removed 1272931 rows and 537430 columns (presolve time = 36s) ...
Presolve removed 1272931 rows and 628630 columns (presolve time = 43s) ...
Presolve removed 1272931 rows and 628630 columns (presolve time = 46s) ...
Presolve removed 1273011 rows and 628630 columns (presolve time = 60s) ...
Presolve removed 1274381 rows and 630000 columns (presolve time = 69s) ...
Presolve removed 1504381 rows and 860000 columns (presolve time = 70s) ...
Presolve removed 2284381 rows and 1640000 columns (presolve time = 75s) ...
Presolve removed 2350108 rows and 1705727 columns (presolve time = 81s) ...
Presolve removed 2350188 rows and 1705807 columns (presolve time = 87s) ...
Presolve removed 2350332 rows and 1705951 columns (presolve time = 138s) ...
Presolve removed 2350332 rows and 1705951 columns (presolve time = 149s) ...
Presolve removed 2350332 rows and 1705951 columns
Presolve time: 151.32s
Presolved: 8270792 rows, 4154905 columns, 31360110 nonzeros
Elapsed ordering time = 5s
Elapsed ordering time = 10s
Elapsed ordering time = 15s
Elapsed ordering time = 20s
Elapsed ordering time = 25s
Ordering time: 25.60s

Barrier statistics:
 Dense cols : 2050
 AA' NZ : 5.782e+07
 Factor NZ : 2.413e+08 (roughly 7.0 GB of memory)
 Factor Ops : 1.319e+11 (roughly 10 seconds per iteration)
 Threads : 28

 Objective Residual
Iter Primal Dual Primal Dual Compl Time
 0 3.41685523e+11 -1.63810339e+07 6.69e+09 1.62e+01 4.81e+07 268s
 1 3.14345117e+11 -8.32337663e+07 6.11e+09 4.25e+01 4.31e+07 291s
 2 2.93727792e+11 -2.55387069e+08 5.67e+09 1.09e+02 3.12e+07 314s
 3 2.46020004e+11 -7.80471364e+08 4.73e+09 3.96e+01 1.50e+07 354s
 4 1.71634745e+11 -2.37836127e+09 3.51e+09 2.05e+01 9.80e+06 385s
 5 6.54088664e+10 -4.54381709e+09 1.48e+09 6.49e+00 4.11e+06 418s
 6 5.82102527e+10 -4.66783548e+09 1.34e+09 5.75e+00 3.70e+06 458s
 7 5.04334016e+10 -5.03433607e+09 1.17e+09 3.55e+00 2.96e+06 491s
 8 3.96289730e+10 -5.39542519e+09 9.20e+08 2.02e+00 2.25e+06 516s
 9 3.47432253e+10 -5.63837852e+09 7.98e+08 1.19e+00 1.91e+06 538s
 10 3.06902181e+10 -5.82068281e+09 6.97e+08 6.83e-01 1.64e+06 554s
 11 2.59532215e+10 -5.83031762e+09 5.77e+08 5.89e-01 1.36e+06 576s
 12 1.95851062e+10 -5.99319714e+09 4.27e+08 1.87e-01 9.46e+05 611s
 13 1.24571408e+10 -6.09529338e+09 2.32e+08 4.51e-02 5.12e+05 644s
 14 4.72464392e+09 -6.15324418e+09 3.69e+07 1.27e-02 8.50e+04 672s
 15 3.50084556e+09 -6.10347632e+09 1.62e+07 5.78e-03 3.83e+04 709s
 16 2.63573718e+09 -5.88615609e+09 8.74e+06 2.59e-03 2.09e+04 733s
 17 1.96630609e+09 -5.62569373e+09 5.07e+06 1.33e-03 1.23e+04 757s
 18 1.12791925e+09 -5.02405482e+09 1.72e+06 8.50e-04 4.47e+03 773s
 19 5.32692309e+08 -3.71723104e+09 5.50e+05 3.60e-04 1.57e+03 793s
 20 2.30754536e+08 -1.81736316e+09 1.52e+05 1.17e-04 4.78e+02 816s
 21 1.08270193e+08 -4.36110806e+08 5.56e+04 4.55e-06 1.39e+02 840s
 22 6.35738248e+07 -2.80876890e+08 2.84e+04 6.28e-07 7.36e+01 858s
 23 4.56594236e+07 -2.06653496e+08 1.80e+04 3.58e-09 4.79e+01 875s
 24 2.03646780e+07 -1.35712185e+08 6.96e+03 2.87e-09 2.27e+01 894s
 25 1.30302019e+07 -6.36789204e+07 3.99e+03 1.29e-09 1.08e+01 916s
 26 7.18426261e+06 -2.95146859e+07 1.96e+03 4.24e-08 4.76e+00 932s
 27 5.39481458e+06 -1.96540188e+07 1.38e+03 5.29e-08 3.11e+00 950s
 28 4.25616871e+06 -1.31168140e+07 1.03e+03 5.25e-08 2.08e+00 972s
 29 3.39122251e+06 -8.56459685e+06 7.87e+02 6.56e-07 1.37e+00 993s
 30 2.35494870e+06 -5.45590145e+06 4.96e+02 4.81e-07 8.34e-01 1014s
 31 1.75961076e+06 -4.32285838e+06 3.35e+02 4.88e-07 6.16e-01 1031s
 32 1.54577811e+06 -3.31696295e+06 2.79e+02 4.44e-07 4.84e-01 1047s
 33 1.24876497e+06 -2.71730748e+06 2.07e+02 4.14e-07 3.84e-01 1067s
 34 9.65563500e+05 -2.10936400e+06 1.42e+02 3.50e-07 2.90e-01 1089s
 35 8.09751695e+05 -1.21789001e+06 1.09e+02 2.71e-07 1.88e-01 1124s
 36 6.85681332e+05 -8.71773859e+05 8.41e+01 2.35e-07 1.42e-01 1159s
 37 6.16762659e+05 -6.59731109e+05 7.17e+01 2.02e-07 1.16e-01 1204s
 38 5.31008170e+05 -5.49450998e+05 5.65e+01 1.99e-07 9.72e-02 1239s
 39 4.85025495e+05 -3.52131013e+05 4.85e+01 1.74e-07 7.49e-02 1282s
 40 4.61067771e+05 -2.86287108e+05 4.47e+01 1.55e-07 6.67e-02 1324s
 41 4.37224810e+05 -2.10862725e+05 4.11e+01 1.29e-07 5.77e-02 1359s
 42 4.27471069e+05 -1.90716208e+05 3.96e+01 1.22e-07 5.50e-02 1395s
 43 4.11232535e+05 -1.38868137e+05 3.72e+01 1.16e-07 4.89e-02 1434s
 44 3.84252057e+05 -9.26527957e+04 3.33e+01 1.07e-07 4.23e-02 1467s
 45 3.75160967e+05 -8.65482286e+04 3.19e+01 1.05e-07 4.09e-02 1494s
 46 3.44869195e+05 -5.10711601e+04 2.77e+01 9.13e-08 3.50e-02 1535s
 47 3.27264089e+05 -3.07525436e+04 2.54e+01 8.17e-08 3.16e-02 1575s
 48 3.16481726e+05 -2.48734815e+04 2.39e+01 7.88e-08 3.01e-02 1609s
 49 3.02984378e+05 -1.42205027e+04 2.20e+01 7.38e-08 2.79e-02 1645s
 50 2.85771137e+05 1.00004760e+04 1.97e+01 6.17e-08 2.43e-02 1690s
 51 2.62876677e+05 3.00995865e+04 1.67e+01 5.15e-08 2.04e-02 1731s
 52 2.53977038e+05 4.57190525e+04 1.56e+01 4.38e-08 1.83e-02 1767s
 53 2.38721422e+05 5.69380032e+04 1.36e+01 3.81e-08 1.59e-02 1804s
 54 2.23758750e+05 6.38469654e+04 1.17e+01 3.47e-08 1.40e-02 1840s
 55 2.11845886e+05 7.60586652e+04 1.01e+01 2.85e-08 1.19e-02 1878s
 56 2.03103042e+05 8.19823726e+04 8.99e+00 2.54e-08 1.06e-02 1918s
 57 2.01927142e+05 8.45401338e+04 8.86e+00 2.39e-08 1.03e-02 1954s
 58 2.00840892e+05 8.61232076e+04 8.72e+00 2.30e-08 1.00e-02 1977s
 59 1.98229044e+05 9.00154969e+04 8.43e+00 2.08e-08 9.45e-03 2008s
 60 1.90612134e+05 9.29345655e+04 7.50e+00 1.92e-08 8.52e-03 2043s
 61 1.86026509e+05 9.70726786e+04 6.95e+00 1.70e-08 7.76e-03 2082s
 62 1.79982287e+05 1.00357567e+05 6.24e+00 1.51e-08 6.94e-03 2114s
 63 1.72656317e+05 1.03675868e+05 5.36e+00 1.32e-08 6.01e-03 2147s
 64 1.68954798e+05 1.06878267e+05 4.92e+00 1.13e-08 5.41e-03 2179s
 65 1.65810746e+05 1.08567916e+05 4.55e+00 1.04e-08 4.99e-03 2220s
 66 1.62776002e+05 1.10648181e+05 4.19e+00 9.21e-09 4.54e-03 2264s
 67 1.59063023e+05 1.12272190e+05 3.76e+00 8.31e-09 4.07e-03 2303s
 68 1.55762553e+05 1.14018578e+05 3.37e+00 7.36e-09 3.63e-03 2343s
 69 1.52942240e+05 1.15322646e+05 3.03e+00 6.66e-09 3.27e-03 2374s
 70 1.51525082e+05 1.16764960e+05 2.85e+00 5.90e-09 3.02e-03 2404s
 71 1.49706038e+05 1.17339374e+05 2.64e+00 5.59e-09 2.82e-03 2444s
 72 1.48559122e+05 1.17957643e+05 2.50e+00 5.26e-09 2.66e-03 2474s
 73 1.47736842e+05 1.18377649e+05 2.40e+00 5.04e-09 2.55e-03 2505s
 74 1.46826390e+05 1.19055775e+05 2.29e+00 4.68e-09 2.42e-03 2537s
 75 1.45675152e+05 1.19550041e+05 2.16e+00 4.41e-09 2.27e-03 2573s
 76 1.45165699e+05 1.19690350e+05 2.09e+00 4.34e-09 2.22e-03 2600s
 77 1.44605514e+05 1.20036785e+05 2.03e+00 4.15e-09 2.14e-03 2628s
 78 1.43542523e+05 1.20765038e+05 1.90e+00 3.75e-09 1.98e-03 2669s
 79 1.42646862e+05 1.21613434e+05 1.80e+00 3.27e-09 1.83e-03 2705s
 80 1.40306892e+05 1.22384352e+05 1.52e+00 2.84e-09 1.56e-03 2747s
 81 1.39037443e+05 1.23126406e+05 1.36e+00 2.44e-09 1.38e-03 2788s
 82 1.37709617e+05 1.23552367e+05 1.20e+00 2.22e-09 1.23e-03 2818s
 83 1.37252174e+05 1.23821418e+05 1.14e+00 2.09e-09 1.17e-03 2847s
 84 1.36402168e+05 1.24259568e+05 1.04e+00 1.86e-09 1.06e-03 2879s
 85 1.35620762e+05 1.24507245e+05 9.44e-01 1.73e-09 9.66e-04 2914s
 86 1.34405665e+05 1.24744023e+05 7.93e-01 1.62e-09 8.39e-04 2954s
 87 1.33793226e+05 1.25048018e+05 7.15e-01 1.46e-09 7.60e-04 2996s
 88 1.33446115e+05 1.25255358e+05 6.72e-01 1.36e-09 7.11e-04 3026s
 89 1.32751939e+05 1.25465545e+05 5.86e-01 1.26e-09 6.33e-04 3063s
 90 1.32399152e+05 1.25915381e+05 5.41e-01 1.04e-09 5.63e-04 3105s
 91 1.32011396e+05 1.26138363e+05 4.92e-01 9.25e-10 5.10e-04 3139s
 92 1.31755439e+05 1.26409233e+05 4.60e-01 7.94e-10 4.64e-04 3171s
 93 1.31373208e+05 1.26617099e+05 4.12e-01 6.93e-10 4.13e-04 3208s
 94 1.31127391e+05 1.26804030e+05 3.82e-01 6.03e-10 3.75e-04 3246s
 95 1.30938594e+05 1.26926921e+05 3.58e-01 5.42e-10 3.48e-04 3275s
 96 1.30556505e+05 1.26992100e+05 3.11e-01 5.11e-10 3.10e-04 3312s
 97 1.30351906e+05 1.27155787e+05 2.84e-01 4.33e-10 2.78e-04 3345s
 98 1.30158862e+05 1.27223615e+05 2.60e-01 4.01e-10 2.55e-04 3375s
 99 1.29988824e+05 1.27368663e+05 2.38e-01 3.33e-10 2.28e-04 3411s
 100 1.29773441e+05 1.27432470e+05 2.11e-01 3.04e-10 2.03e-04 3451s
 101 1.29664925e+05 1.27511857e+05 1.98e-01 2.91e-10 1.87e-04 3488s
 102 1.29547706e+05 1.27540128e+05 1.83e-01 2.88e-10 1.74e-04 3522s
 103 1.29445251e+05 1.27575778e+05 1.70e-01 2.85e-10 1.62e-04 3558s
 104 1.29331770e+05 1.27612230e+05 1.56e-01 2.80e-10 1.49e-04 3592s
 105 1.29288047e+05 1.27648602e+05 1.50e-01 2.72e-10 1.42e-04 3622s
 106 1.29183819e+05 1.27688982e+05 1.37e-01 2.68e-10 1.30e-04 3656s
 107 1.29106075e+05 1.27712668e+05 1.27e-01 2.63e-10 1.21e-04 3689s
 108 1.29047034e+05 1.27742895e+05 1.20e-01 2.57e-10 1.13e-04 3733s
 109 1.28936232e+05 1.27768894e+05 1.06e-01 2.51e-10 1.01e-04 3771s
 110 1.28828363e+05 1.27796713e+05 9.21e-02 2.40e-10 8.96e-05 3805s
 111 1.28786040e+05 1.27820070e+05 8.66e-02 2.32e-10 8.39e-05 3843s
 112 1.28736737e+05 1.27843844e+05 8.03e-02 2.22e-10 7.75e-05 3881s
 113 1.28712806e+05 1.27857136e+05 7.73e-02 2.18e-10 7.43e-05 3903s
 114 1.28692018e+05 1.27863362e+05 7.46e-02 2.16e-10 7.19e-05 3932s
 115 1.28646428e+05 1.27881673e+05 6.90e-02 2.09e-10 6.64e-05 3964s
 116 1.28601905e+05 1.27918654e+05 6.32e-02 1.94e-10 5.93e-05 4003s
 117 1.28556881e+05 1.27928834e+05 5.75e-02 1.90e-10 5.45e-05 4042s
 118 1.28532679e+05 1.27941298e+05 5.44e-02 1.84e-10 5.13e-05 4079s
 119 1.28501672e+05 1.27960624e+05 5.05e-02 1.77e-10 4.70e-05 4121s
 120 1.28477468e+05 1.27970454e+05 4.74e-02 1.73e-10 4.40e-05 4149s
 121 1.28447448e+05 1.27981110e+05 4.37e-02 1.69e-10 4.05e-05 4185s
 122 1.28443690e+05 1.27983510e+05 4.33e-02 1.68e-10 4.00e-05 4221s
 123 1.28433742e+05 1.27989497e+05 4.20e-02 1.67e-10 3.86e-05 4244s
 124 1.28410909e+05 1.27995229e+05 3.91e-02 1.64e-10 3.61e-05 4276s
 125 1.28380678e+05 1.28004977e+05 3.53e-02 1.61e-10 3.26e-05 4309s
 126 1.28369696e+05 1.28012703e+05 3.39e-02 1.58e-10 3.10e-05 4338s
 127 1.28361750e+05 1.28015962e+05 3.29e-02 1.57e-10 3.00e-05 4370s
 128 1.28334753e+05 1.28026805e+05 2.96e-02 1.55e-10 2.67e-05 4406s
 129 1.28316213e+05 1.28038230e+05 2.73e-02 1.52e-10 2.41e-05 4439s
 130 1.28295537e+05 1.28043262e+05 2.48e-02 1.51e-10 2.19e-05 4477s
 131 1.28285135e+05 1.28050030e+05 2.36e-02 1.54e-10 2.04e-05 4520s
 132 1.28278168e+05 1.28053589e+05 2.27e-02 1.46e-10 1.95e-05 4559s
 133 1.28262631e+05 1.28058866e+05 2.08e-02 1.42e-10 1.77e-05 4600s
 134 1.28254405e+05 1.28062150e+05 1.97e-02 1.39e-10 1.67e-05 4635s
 135 1.28240428e+05 1.28064506e+05 1.80e-02 1.36e-10 1.53e-05 4668s
 136 1.28232110e+05 1.28067062e+05 1.70e-02 1.33e-10 1.43e-05 4706s
 137 1.28221892e+05 1.28069327e+05 1.57e-02 1.29e-10 1.32e-05 4740s
 138 1.28210338e+05 1.28072626e+05 1.42e-02 1.50e-10 1.20e-05 4775s
 139 1.28208093e+05 1.28076060e+05 1.39e-02 1.86e-10 1.15e-05 4812s
 140 1.28190223e+05 1.28077644e+05 1.17e-02 2.28e-10 9.78e-06 4850s
 141 1.28187251e+05 1.28078407e+05 1.13e-02 2.46e-10 9.45e-06 4878s
 142 1.28184067e+05 1.28080956e+05 1.09e-02 3.29e-10 8.96e-06 4915s
 143 1.28170629e+05 1.28082301e+05 9.24e-03 3.45e-10 7.67e-06 4949s
 144 1.28165990e+05 1.28083802e+05 8.65e-03 3.68e-10 7.14e-06 4986s
 145 1.28157101e+05 1.28084814e+05 7.52e-03 3.71e-10 6.28e-06 5019s
 146 1.28151511e+05 1.28085523e+05 6.80e-03 3.70e-10 5.73e-06 5055s
 147 1.28146842e+05 1.28086975e+05 6.20e-03 3.60e-10 5.20e-06 5089s
 148 1.28144150e+05 1.28088251e+05 5.86e-03 3.58e-10 4.85e-06 5129s
 149 1.28139076e+05 1.28089349e+05 5.23e-03 3.52e-10 4.32e-06 5165s
 150 1.28136184e+05 1.28090200e+05 4.86e-03 3.37e-10 3.99e-06 5210s
 151 1.28132938e+05 1.28091231e+05 4.46e-03 3.16e-10 3.62e-06 5247s
 152 1.28130240e+05 1.28091764e+05 4.12e-03 3.04e-10 3.34e-06 5285s
 153 1.28127164e+05 1.28091951e+05 3.72e-03 3.00e-10 3.06e-06 5314s
 154 1.28125922e+05 1.28093081e+05 3.56e-03 2.72e-10 2.85e-06 5349s
 155 1.28123672e+05 1.28093309e+05 3.28e-03 2.63e-10 2.64e-06 5383s
 156 1.28121304e+05 1.28093759e+05 2.97e-03 2.46e-10 2.39e-06 5418s
 157 1.28118771e+05 1.28094094e+05 2.65e-03 2.56e-10 2.14e-06 5452s
 158 1.28116394e+05 1.28094380e+05 2.36e-03 2.64e-10 1.91e-06 5486s
 159 1.28114146e+05 1.28094568e+05 2.07e-03 2.70e-10 1.70e-06 5526s
 160 1.28112375e+05 1.28094782e+05 1.84e-03 2.76e-10 1.53e-06 5562s
 161 1.28111291e+05 1.28095224e+05 1.71e-03 2.90e-10 1.40e-06 5603s
 162 1.28109803e+05 1.28095409e+05 1.52e-03 2.96e-10 1.25e-06 5642s
 163 1.28109330e+05 1.28095691e+05 1.46e-03 3.06e-10 1.18e-06 5679s
 164 1.28108348e+05 1.28095768e+05 1.34e-03 3.09e-10 1.09e-06 5708s
 165 1.28106863e+05 1.28096070e+05 1.15e-03 3.20e-10 9.37e-07 5745s
 166 1.28105814e+05 1.28096234e+05 1.02e-03 3.26e-10 8.32e-07 5787s
 167 1.28104981e+05 1.28096395e+05 9.16e-04 3.32e-10 7.46e-07 5825s
 168 1.28104094e+05 1.28096523e+05 8.02e-04 3.38e-10 6.58e-07 5862s
 169 1.28103226e+05 1.28096687e+05 6.91e-04 3.45e-10 5.68e-07 5895s
 170 1.28102525e+05 1.28096908e+05 6.01e-04 3.54e-10 4.88e-07 5923s
 171 1.28101990e+05 1.28096973e+05 5.34e-04 3.57e-10 4.36e-07 5968s
 172 1.28101576e+05 1.28097121e+05 4.82e-04 3.65e-10 3.87e-07 6005s
 173 1.28101351e+05 1.28097236e+05 4.54e-04 3.71e-10 3.57e-07 6039s
 174 1.28100561e+05 1.28097281e+05 3.56e-04 3.73e-10 2.85e-07 6078s
 175 1.28100138e+05 1.28097345e+05 3.03e-04 3.77e-10 2.43e-07 6117s
 176 1.28099881e+05 1.28097401e+05 2.72e-04 3.80e-10 2.15e-07 6159s
 177 1.28099624e+05 1.28097434e+05 2.40e-04 3.82e-10 1.90e-07 6204s
 178 1.28099575e+05 1.28097448e+05 2.34e-04 3.83e-10 1.85e-07 6259s
 179 1.28099241e+05 1.28097476e+05 1.91e-04 3.85e-10 1.53e-07 6295s
 180 1.28099032e+05 1.28097497e+05 1.64e-04 3.87e-10 1.33e-07 6345s
 181 1.28098844e+05 1.28097519e+05 1.41e-04 3.88e-10 1.15e-07 6388s
 182 1.28098703e+05 1.28097546e+05 1.24e-04 3.90e-10 1.01e-07 6440s
 183 1.28098545e+05 1.28097571e+05 1.04e-04 3.92e-10 8.46e-08 6469s
 184 1.28098475e+05 1.28097583e+05 9.48e-05 3.93e-10 7.74e-08 6503s
 185 1.28098368e+05 1.28097588e+05 8.15e-05 3.93e-10 6.78e-08 6551s
 186 1.28098285e+05 1.28097614e+05 7.09e-05 3.95e-10 5.82e-08 6602s
 187 1.28098153e+05 1.28097633e+05 5.43e-05 3.97e-10 4.52e-08 6644s
```

RuaridhMacd commented:
@Betristor I have some ideas for speeding up the writing. Is the OneZone example in your branch representative of the problem? I want to test it on a small case so I can try it multiple times.

RuaridhMacd commented:
@Betristor I tried some quick improvements to the writing functions, but I only had time to test the ThreeZone case. Let me know if they help with larger problems. 

I think there's more we could improve, but perhaps should move it to another PR.

Betristor commented:
@RuaridhMacd the threezones case is quite fast either for establishing the model or accessing results. I faked a ten zone test case based on threezones case. Anna's case is more appropriate for testing your improvements since it has 76 zones and took nearly 10 hours to finish. I will attach them in email to you if you would like to experiment with the improvements.

RuaridhMacd commented:
- [ ] @RuaridhMacd Merge the existing code
- [ ] @RuaridhMacd email @Betristor to understand the issue with N>70 truck routes

RuaridhMacd commented:
- [ ] @RuaridhMacd Review with large model
- [ ] @Betristor Add figure to the documentation which explains the structure of the truck module. Use IEEE paper for reference.

Betristor commented:
@RuaridhMacd I've added the figure to present the truck scheduling model and added detailed description about variables used in the model with more explanation about the constraints.

RuaridhMacd commented:
- [ ] @gn-he Please review the new documentation, then merge if you're happy with it.
